### ISSUE TITLE
Unable to create SSL context

### ISSUE THREAD

**Issue by keshavakarthik** [IBM-Swift/Kitura#959]
With  Kitura (.Package(url: "https://github.com/IBM-Swift/Kitura", majorVersion: 1, minor: 4)) adding SSL Support ( Self-Signed ) giving the error
"Error accepting client connection: Error code: 12, ERROR: Unable to create SSL context."

Whereas the same code is working fine with Kitura Version (Package(url: "https://github.com/IBM-Swift/Kitura.git", majorVersion: 1, minor: 0))

**Comment by na-gupta** [comment]
@gtaban @shmuelk 
It looks like this problem was originally introduced with IBM-Swift/Kitura-net#155 in [this commit](https://github.com/IBM-Swift/Kitura-net/commit/93ddb4126ae6555b1856c8dda96e40d197216e16) between Kitura-net 1.3.0 and 1.3.1

On that commit, the problem goes away with commenting out [this line](https://github.com/IBM-Swift/Kitura-net/blob/master/Sources/KituraNet/IncomingSocketHandler.swift#L313) that was added in that commit. However, commenting out that same line on master does not fix the issue. So it looks like some other changes since then makes the problem reappear.

Further investigation is needed to find out how this should be properly fixed.

**Comment by shmuelk** [comment]
@navneet thanks for the research work.

@billabt I did an additional test here, after you pointed out to me that SSL on Kitura 1.2 works and fails on later versions. I commented out line 1305 in Socket.swift (self.delegate?.deinitialize() ). This is in the function deinit of Socket.swift. After commenting out that line of code SSL **works**.

I believe the problem is in the deinitialization of SLLService instances somehow, or somehow they are being shared across Socket instances.

**Comment by billabt** [comment]
Fixed in BlueSocket v0.12.18 and BlueSSLService v0.12.12.

**Comment by na-gupta** [comment]
@billabt Thanks. That fixes the problem for the most part. But in some [new unit tests](https://github.com/IBM-Swift/Kitura/blob/issue_962/Tests/KituraTests/TestStaticFileServer.swift#L58-L60) I added in [branch 962](https://github.com/IBM-Swift/Kitura/tree/issue_962), there is a new failure:
```
xctest(6295,0x70000666b000) malloc: *** error for object 0x7fb5c9653af0: pointer being freed was not allocated
*** set a breakpoint in malloc_error_break to debug
```
It seems to occur in this line: `let status: OSStatus = SSLWrite(sslContext, buffer, bufSize, &processed)` in SSLService. Could you take a look to see if there's something wrong with the test or some other lingering problem?

**Comment by billabt** [comment]
Did you set a breakpoint in malloc_error_break?  If not, can you try the test again with that breakpoint set and then post the stack trace?

**Comment by billabt** [comment]
There's actually no freeing of buffers taking place anywhere in this section of code.  It's possible that it's happening inside SSLWrite itself or just prior to it...  The stack trace should give us a clue.

**Comment by na-gupta** [comment]
@billabt Here is the stack trace:
<img width="762" alt="screen shot 2017-01-11 at 10 59 37 am" src="https://cloud.githubusercontent.com/assets/12917107/21858069/52e9a408-d7ed-11e6-9c0b-294f012fa056.png">

**Comment by na-gupta** [comment]
@billabt Turns out this was caused by a bug in Kitura where it was calling Socket.write() after calling Socket.close() in some circumstances. I will put in a fix for Kitura. It may be worthwhile to put in a check for this in Socket and/or SSLService as well.

**Comment by na-gupta** [comment]
@keshavakarthik Please try again with the fixes in Kitura 1.6
